### DISCUSSION TITLE
Object names

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#64] [Ideas]
I have been wondering whether we should allow use of object names in expressions. Currently we only allow variables, as in the demos.

Should the user be able to say `(add-shape 'my-scene 'polygon-23)`?

If we do allow this, then we must ensure that object names are unique. What is a good way of doing this?

We also need fast object name lookup, and intuitive object name generation and duplication schemes.

**Comment by mikelevins** [comment]
SK8 had this. It's extremely handy for tinkering in the repl. It's a huge pain in the ass to have a pile of special global names around, though. I'm sort of mildly against it for the latter reason, but not enough to argue with people who really want it. I have to admit that it's really nice for interactive experimentation.

**Comment by kaveh808** [comment]
I'll have a think about this, as the consequences are as you say rather severe.

The meantime, I would like to consider the issue of name generation. I currently have a global counter which is used increment a number for the names of every instance created. So `polygon-1`, `transform-2`, `polyhedron-3`, etc.

Is there a way of having a per-class counter? I was not able to think of a scheme for that.

**Reply by mikelevins** [reply]
sk8 did pretty much exactly this:

```
(format nil "~A-~A" 
  (type-of thing)
  (gethash (type-of thing) *galactic-type-to-counter-map*))
```

That's not literally the sk8 code; just an off-the-top-of-the-head reconstruction of what it was doing. Sk8 interned these and used them as the names of special variables, though that itn't the only way you could use them.

**Reply by loskool** [reply]
Sure you can do a per-class counter by either using a class-level slot (`:allocation class` in your `defclass`) and generating the object name inside the implementation of `initialize-instance :after` or, if you want to get more zany, you can create a metaclass that gives its instance classes a special counter value that is incremented on a specialization to `make-instance`.  That way, every time you make a new class with the appropriate metaclass, the "name generation" logic will be built in. 

You could make a specific package, called something like `kons-9-object-names` and generate all object names in there, so as not to pollute the other packages. If you expect that you'll have thousands of names, however, using a symbol and `symbol-value` or `(setf (get ..))` might not be best. 

Instead, you can use a hashtable with string keys. The table can be stored as a class level slot on a superclass from which all other geometry classes inherit. Or, if you prefer, the table can be bound to a global. After an instance is initialized, it can be inserted into that table by the auto-generated name.

**Reply by kaveh808** [reply]
> Instead, you can use a hashtable with string keys. The table can be stored as a class level slot on a superclass from which all other geometry classes inherit. Or, if you prefer, the table can be bound to a global. After an instance is initialized, it can be inserted into that table by the auto-generated name.

I could not figure out how to make the naming work using class slots. A subclass would still update the superclasses slot value, no?

Are string and symbol keys the same speed for hash tables? I can use EQ for symbol comparison, but I'm not sure if there is an implicit conversion to strings each time.

**Reply by loskool** [reply]
re: speed. No you're correct, symbol lookup should be faster. 

re: per-class counter. something like this:

``` lisp


(defclass named (standard-class)
  ((counter :initform 0 :accessor class-counter))
  (:documentation "This is our metaclass. Its instances are
  classes. These classes will all have a COUNTER slot on them."))

;; this is boiler plate to let CLOS know that we know what we're doing.
(defmethod closer-mop:validate-superclass  
    ((sub named) (sup standard-class))  
  t)


;; this is a super class that will hold a hash table.
(defclass named-instance ()
  ((name :accessor instance-name :documentation "a symbol naming the instance")))

(defvar *table* (make-hash-table))

(defmethod make-instance :around ((class named) &key)
  (let ((instance
          (call-next-method))
        (idx
          (incf (class-counter class))))
    (when (typep instance 'named-instance)
      (setf (instance-name instance)
            (intern (format nil "~a-~a"
                            (class-name class)
                            idx))
            (gethash (instance-name instance) *table*)
            instance))
    instance))

(defclass point (named-instance)
  ((x :initform 0 :accessor x :initarg :x)
   (y :initform 0 :accessor y :initarg :y))
  (:metaclass named))


(defclass polygon (named-instance)
  ((path :initform (list) :documentation "a list of points"))
  (:metaclass named))


(defun lookup-by-name (symbol)
  (gethash symbol *table*))
```

**Reply by loskool** [reply]
I realize that the above example could be modified to keep the table in the metaclass, if so desired.

**Reply by mikelevins** [reply]
You can define a class that contains a counter in a class-allocated slot, and define a subclass that redeclares the same-named slot, also with class allocation. The two classes will then have separate class-allocated slots of the same name. Incrementing the counter in the subclass will not increment it in the supecrclass, and vice versa. I tested in SBCL just to make sure.

**Reply by kaveh808** [reply]
I went simple old-school and did the following. I like that it keeps the naming solely in one class.

```
(defparameter *item-name-table* (make-hash-table :test 'eq))

(defun next-item-type-number (type)
  (let ((num (gethash type *item-name-table*)))
    (if (null num)
        (progn
          (setf (gethash type *item-name-table*) 2)
          1)
        (progn
          (incf (gethash type *item-name-table*))
          num))))

;;; root class for scene entities
(defclass item ()
  ((name :accessor name :initarg :name :initform nil)))

(defmethod initialize-instance :after ((item item) &rest initargs)
  (declare (ignore initargs))
  (when (null (name item))
    (let ((base-name (class-name (class-of item))))
      (setf (name item) (mashup-symbol base-name '- (next-item-type-number base-name))))))
```

**Reply by loskool** [reply]
One thing that I'm not quite sure about is why does it matter that the index counters are distinct? What's wrong with `polygon-2` coming directly before `polyhedra-3`?  Are the numbers themselves going to be used for anything other making symbols to name objects? Do whole classes of instances need to be iterated over or something?

**Reply by mikelevins** [reply]
@loskool, that seems like a fair question.

**Reply by kaveh808** [reply]
User interface concerns. When a user creates an object using the UI, it is more sensible to make sphere-1 than sphere-3625. The object name will appear in the various UI components.

**Reply by loskool** [reply]
ah-hah cool cool

**Reply by mikelevins** [reply]
@kaveh808 that makes sense

**Reply by kaveh808** [reply]
Which brings up the next question. If the user is going to use the object names to get at the object, we need to ensure the object name is unique.

This does mean keeping a record of every object created and making sure we do not duplicate names.

**Reply by mikelevins** [reply]
I think it does mean that, yes, but do we mean unique per session or globally unique? I mean, if we want object names to be unique within a running image, then we can't restart the counter after a save-lisp-and-die and a restart from the image. The image must preserve the states of any counters, or else we could generate name collisions. Which, in the long run, might make the notion of sphere-1 versus sphere-3625 moot.

**Reply by mikelevins** [reply]
Oh, and it gets messy if you want to support either network collaboration or concurrent updates from more than one device, because then you're potentially running more than one counter on more than one device, and you can get collisions that way.

**Reply by kaveh808** [reply]
Assuming we take a scene-based approach, and I'm not sure yet if that's a good idea, object names will only matter when an object is in the scene. That would be the only time the object name would appear in the UI. Therefore if the user deletes a scene, or saves it and starts a new one, all object names can be reset.

But I need to think more about this.

**Comment by JMC-design** [comment]
I'm not sure why there'd be any need for specials.  Objects already have symbol names, throw in a typecase and if a symbol (find symbol-name scene-shapes :key object-name).   though to my thinking that all belongs in some sort of heirarchy api.

**Reply by kaveh808** [reply]
We do currently have `find-shape-by-name`.

I thought of the typecase usage, but then we would have two classes of functions: those that understand symbols as objects and those that don't. It would divide things into an API and just raw Lisp code. A distinction I don't want to make, as one of our main selling points is that the code is the API.

**Comment by mikelevins** [comment]
"I'm not sure why there'd be any need for specials."

I'm not talking about special variables. What I mean is that if you're generating globally-accessible symbols to refer to every object you interact with, you end up with hundreds of thousands of random symbols floating around the image, and that's not without consequence.

Now you *could* also make them special variables, and sk8 did, because that's also pretty handy in the repl. For example, in sk8 you could grab any arbitrary object on the screen (a rectangle, a movie, a window, a close button from a window), drag it over a repl window and drop it, and you'd get a special variable that referred to the dragged object. This is incredibly handy for tinkering with stuff. That generated special variable is a reference you can use to do anything you like with the object you dropped on the repl window.

But housekeeping of hundreds of thousands of autogenerated symbols remains an issue.

**Comment by JMC-design** [comment]
Well all those symbols would be confined to whatever package the person is developing in.
I do think objects should have an integer id, which is how you'd refer to them in core, and that object names shouldn't be autogenerated but meaningful to the content creator, e.g. "red teapot" or arkanoid-assets:brick

**Reply by kaveh808** [reply]
Users can specify object names, but you still need to generate a default name if they don't.

**Reply by mikelevins** [reply]
Autogenerating names is extremely convenient, as exemplified, again, by sk8. Take the example of grabbing an arbitrary object from an arbitrary scene and dropping it on the repl. sk8 would autocreate a variable that referred to the dropped object. That's just far handier than any other way I've seen to get references to arbitrary objects.

If the only way objects get names is that the people who construct the objects assign them, that just means that most objects don't have names.

**Comment by mikelevins** [comment]
So if object names are interned symbols and they are autogenerated, then a user just interacting normally with the system is going to constantly be generating unlimited numbers of new symbols and interning them. That's less than ideal.

If they're interned symbols, then they can be used as the names of special variables, and then you can have sk8-like convenience in referencing arbitrary objects. But maybe the cost is too high.

An alternative would be to name objects with strings or uninterned symbols, and use a find-object function, analogous to find-class, that looks up objects in a table.

The table has sort of the same problem as whatever package autogenerated symbols would have gone into, in that it gets stuffed with endless numbers of generated strings, but sbcl has an extension to hash tables that allows them to be weak: if nothing but the hash table refers to a key, then GC can discard it. So the table of object keys doesn't have to grow endlessly without bound.

**Reply by loskool** [reply]
+1 on hashtables with weak refs

**Reply by kaveh808** [reply]
I will continue with the current approach of uninterned symbols for object names. The lookup functions `find-shape-by-name` and `find-motion-by-name` require a scene for the lookup. This means garbage collection should not be an issue as we only find objects in a given scene. If we need more general functionality in the future, we can use weak hash tables.
